pyflakes/checker.py

```diff
@@ -518,6 +527,60 @@ def is_typing_overload_decorator(node):
     )
 
 
+def make_tokens(code):
+    # PY3: tokenize.tokenize requires readline of bytes
+    if not isinstance(code, bytes):
+        code = code.encode('UTF-8')
+    lines = iter(code.splitlines(True))
+    # next(lines, b'') is to prevent an error in pypy3
+    return tuple(tokenize_tokenize(lambda: next(lines, b'')))
+
+
+class _TypeableVisitor(ast.NodeVisitor):
+    """Collect the line number and nodes which are deemed typeable by
+    PEP 484
+
+    https://www.python.org/dev/peps/pep-0484/#type-comments
+    """
+    def __init__(self):
+        self.typeable_lines = []  # type: int
+        self.typeable_nodes = {}  # type: Dict[int, ast.AST]
+
+    def _typeable(self, node):
+        # if there is more than one typeable thing on a line last one wins
+        self.typeable_lines.append(node.lineno)
+        self.typeable_nodes[node.lineno] = node
+
+        self.generic_visit(node)
+
+    visit_Assign = visit_For = visit_FunctionDef = visit_With = _typeable
+    visit_AsyncFor = visit_AsyncFunctionDef = visit_AsyncWith = _typeable
+
+
+def _collect_type_comments(tree, tokens):
+    visitor = _TypeableVisitor()
+    visitor.visit(tree)
+
+    type_comments = collections.defaultdict(list)
+    for tp, text, start, _, _ in tokens:
+        if (
+                tp != tokenize.COMMENT or  # skip non comments
+                not text.startswith('# type:') or  # skip non-type comments
+                text.split()[:3] == ['#', 'type:', 'ignore']  # skip ignores
```

Comment:
Perhaps another silly question, why are we doing `text.split()[:3]` to check this but not using split above to see if the comment starts with `# type:`?

My guess is that the split here is to accommodate multiple spaces so that all of the following work:

```
# type: ignore
#   type:       ignore
# type:    ignore
#      type: ignore
```

But then we're going to bail on those lines anyway because of `not text.startswith('# type:')` so I'd like to understand the decision here. Do we perhaps want to do:

```py
for tp, text, start, _, _ in tokens:
    text_pieces = text.split()
    if (
        tp != tokenize.COMMENT or
        not (text_pieces[:2] == ['#', 'type:'] and
             text_pieces[3] == 'ignore')
    ):
        comment
```

Reply:
I don't know the intricacies of mypy but does it support `#type:ignore` as well? Do we need to handle that? Should this be a regular expression?

Reply:
yeah looks like there's more to support here, I'll make a regex for this.  Even this is valid: `#type:()->None`